### ISSUE TITLE
tensorrt does not generate the correct yolo11 model

### ISSUE THREAD

**Issue by Mshir0** [marcoslucianops/DeepStream-Yolo#592]
The device I'm using is a Jetson Xavier NX. 
Ubuntu version is 20.04
The jetpack version is 5.1.4
cuda version 11.4
cudnn version is 8.6.0.166
tensorrt version 8.5.2.2
Related libraries for generating .onnx
onnx 1.17.0
onnxsim 0.4.36
onnxslim 0.1.39
onnxrumtime-gpu 1.18.0
Then the following are the runtime warnings:
```
nvidia@nvidia:~/DeepStream-Yolo$ deepstream-app -c deepstream_app_config.txt 
WARNING: Deserialize engine failed because file path: /home/nvidia/DeepStream-Yolo/yolo11n.pt.onnx_b1_gpu0_fp16.plan open error
0:00:04.696507296 1030812 0xaaaad9b3eab0 WARN                 nvinfer gstnvinfer.cpp:679:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1976> [UID = 1]: deserialize engine from file :/home/nvidia/DeepStream-Yolo/yolo11n.pt.onnx_b1_gpu0_fp16.plan failed
0:00:04.777985167 1030812 0xaaaad9b3eab0 WARN                 nvinfer gstnvinfer.cpp:679:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:2081> [UID = 1]: deserialize backend context from engine from file :/home/nvidia/DeepStream-Yolo/yolo11n.pt.onnx_b1_gpu0_fp16.plan failed, try rebuild
0:00:04.778568899 1030812 0xaaaad9b3eab0 INFO                 nvinfer gstnvinfer.cpp:682:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:2002> [UID = 1]: Trying to create engine from model files
WARNING: [TRT]: onnx2trt_utils.cpp:375: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
WARNING: [TRT]: onnx2trt_utils.cpp:403: One or more weights outside the range of INT32 was clamped
WARNING: [TRT]: Tensor DataType is determined at build time for tensors not marked as input or output.

Building the TensorRT Engine

6.77WARNING: [TRT]: TensorRT encountered issues when converting weights between types and that could affect accuracy.
WARNING: [TRT]: If this is not the desired behavior, please modify the weights or retrain with regularization to adjust the magnitude of the weights.
WARNING: [TRT]: Check verbose logs for the list of affected weights.
WARNING: [TRT]: - 2 weights are affected by this issue: Detected NaN values and converted them to corresponding FP16 NaN.
WARNING: [TRT]: - 75 weights are affected by this issue: Detected subnormal FP16 values.
WARNING: [TRT]: - 1 weights are affected by this issue: Detected values less than smallest positive FP16 subnormal value and converted them to the FP16 minimum subnormalized value.
Building complete

0:23:06.223452350 1030812 0xaaaad9b3eab0 INFO                 nvinfer gstnvinfer.cpp:682:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:2034> [UID = 1]: serialize cuda engine to file: /home/nvidia/DeepStream-Yolo/model_b1_gpu0_fp16.engine successfully
WARNING: [TRT]: The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
INFO: [Implicit Engine Info]: layers num: 2
0   INPUT  kFLOAT input           3x640x640       
1   OUTPUT kFLOAT output          8400x6          

0:23:06.362116421 1030812 0xaaaad9b3eab0 INFO                 nvinfer gstnvinfer_impl.cpp:328:notifyLoadModelStatus:<primary_gie> [UID 1]: Load new model:/home/nvidia/DeepStream-Yolo/config_infer_primary_yoloV8.txt sucessfully
```
I can found the video stream but the .engine model doesn't detect anything.

**Comment by Mshir0** [comment]
I am using export_yoloV8.py to generate the ONNX model, and here is how I created the model.
'''
python3 export_yoloV8.py -w yolo11n.pt --simplify
'''

**Comment by Mshir0** [comment]
I am using export_yoloV8.py to generate the ONNX model, and here is how I created the model.
```
python3 export_yoloV8.py -w yolo11n.pt --simplify
```

**Comment by Mshir0** [comment]
To add, the version of deepsteam I'm using is 6.3 and I can run deepstream-test1-app successfully.

**Comment by Sanelembuli98** [comment]
I am also having the same problem having the same problem with yolo v5. I resorted to modifying the export_yoloV5.py as below import os
import onnx
import torch
import torch.nn as nn
from models.experimental import attempt_load


class DeepStreamOutput(nn.Module):
    def __init__(self, split_outputs=False):
        """
        :param split_outputs: If True, the forward method will return separate tensors
                              (boxes, scores, classes) instead of a concatenated tensor.
        """
        super().__init__()
        self.split_outputs = split_outputs

    # def forward(self, x):
    #     x = x[0]  # Assuming the first tensor in the list is the output
    #     boxes = x[:, :, :4]
    #     convert_matrix = torch.tensor(
    #         [[1, 0, 1, 0], [0, 1, 0, 1], [-0.5, 0, 0.5, 0], [0, -0.5, 0, 0.5]],
    #         dtype=boxes.dtype, device=boxes.device
    #     )
    #     boxes @= convert_matrix
    #     objectness = x[:, :, 4:5]
    #     scores, labels = torch.max(x[:, :, 5:], dim=-1, keepdim=True)
    #     scores *= objectness

    #     if self.split_outputs:
    #         return boxes, scores, labels  # Return separate tensors
    #         #return boxes, scores, labels.to(boxes.dtype)  # Return separate tensors
    #         #return torch.stack([boxes, scores, labels.to(boxes.dtype)], dim=0)
    #     else:
    #         return torch.cat([boxes, scores, labels.to(boxes.dtype)], dim=-1)  # Return concatenated tensor

    def forward(self, x):
        x = x[0]  # Assuming the first tensor in the list is the output
        boxes = x[:, :, :4]
        
        # Define the conversion matrix for box adjustments
        convert_matrix = torch.tensor(
            [[1, 0, -0.5, 0],
        [0, 1, 0, -0.5],
        [1, 0, 0.5, 0],
        [0, 1, 0, 0.5]], 
        dtype=boxes.dtype, device=boxes.device
        )
        boxes @= convert_matrix
        # Extract objectness scores and class scores
        objectness = x[:, :, 4:5]  # Shape: [batch_size, num_boxes, 1]
        scores, labels = torch.max(x[:, :, 5:], dim=-1, keepdim=True)  # Shape: [batch_size, num_boxes, 1]
        
        # Adjust scores using objectness
        scores = scores * objectness

        print(f"Boxes shape: {boxes.shape}, Scores shape: {scores.shape}, Labels shape: {labels.shape}")

        if self.split_outputs:
            # Directly return boxes, scores, and labels for separate outputs
            return boxes, scores, labels
        
        else:
            # Ensure labels are converted to match the dtype of `boxes` for concatenation
            labels = labels.to(dtype=boxes.dtype)  # Convert labels to match the dtype of boxes
            return torch.cat([boxes, scores, labels], dim=-1)  # Concatenate along the last dimension


def parse_tensorrt_output(output):
    """
    Parses TensorRT output, assuming the shape is [25200, 6].
    :param output: Tensor with shape [25200, 6].
    :return: boxes, confidence scores, and class predictions as separate tensors.
    """
    boxes = output[:, :4]
    confidence = output[:, 4:5]
    classes = output[:, 5:]
    return boxes, confidence, classes


def yolov5_export(weights, device, inplace=True, fuse=True):
    model = attempt_load(weights, device=device, inplace=inplace, fuse=fuse)
    model.eval()
    for k, m in model.named_modules():
        if m.__class__.__name__ == 'Detect':
            m.inplace = False
            m.dynamic = False
            m.export = True
    return model


def suppress_warnings():
    import warnings
    warnings.filterwarnings('ignore', category=torch.jit.TracerWarning)
    warnings.filterwarnings('ignore', category=UserWarning)
    warnings.filterwarnings('ignore', category=DeprecationWarning)
    warnings.filterwarnings('ignore', category=FutureWarning)
    warnings.filterwarnings('ignore', category=ResourceWarning)


def validate_onnx(onnx_file):
    import onnx
    try:
        print('Validating the ONNX model...')
        if isinstance(onnx_file, str):  # If it's a file path, load it
            model = onnx.load(onnx_file)
        else:  # Assume it's already a ModelProto object
            model = onnx_file
        onnx.checker.check_model(model)  # Validate the loaded model
        print('ONNX model validation: Success!')
    except Exception as e:  # Catch all validation-related errors
        print(f'ONNX model validation failed: {e}')


def main(args):
    suppress_warnings()

    print(f'\nStarting: {args.weights}')

    print('Opening YOLOv5 model')
    device = torch.device('cpu')
    model = yolov5_export(args.weights, device)

    if len(model.names.keys()) > 0:
        print('Creating labels.txt file')
        with open('labels.txt', 'w', encoding='utf-8') as f:
            for name in model.names.values():
                f.write(f'{name}\n')

    # Set split_outputs to True if you need separate outputs
    model = nn.Sequential(model, DeepStreamOutput(split_outputs=args.split_outputs))

    img_size = args.size * 2 if len(args.size) == 1 else args.size
    if img_size == [640, 640] and args.p6:
        img_size = [1280] * 2

    onnx_input_im = torch.zeros(args.batch, 3, *img_size).to(device)
    onnx_output_file = f'{args.weights}.onnx'

    dynamic_axes = {'input': {0: 'batch'}}
    if args.split_outputs:
        dynamic_axes.update({'boxes': {0: 'batch'}, 'scores': {0: 'batch'}, 'classes': {0: 'batch'}})
        output_names = ['boxes', 'scores', 'classes']
    else:
        dynamic_axes.update({'output': {0: 'batch'}})
        output_names = ['output']

    print('Exporting the model to ONNX')
    torch.onnx.export(
        model, onnx_input_im, onnx_output_file, verbose=False, opset_version=args.opset, do_constant_folding=True,
        input_names=['input'], output_names=output_names,
        dynamic_axes=dynamic_axes if args.dynamic else None
    )

    if args.simplify:
        print('Simplifying the ONNX model')
        import onnxslim
        model_onnx = onnx.load(onnx_output_file)
        model_onnx = onnxslim.slim(model_onnx)
        validate_onnx(model_onnx)
        onnx.save(model_onnx, onnx_output_file)

    print(f'Done: {onnx_output_file}\n')


def parse_args():
    import argparse
    parser = argparse.ArgumentParser(description='DeepStream YOLOv5 conversion')
    parser.add_argument('-w', '--weights', required=True, type=str, help='Input weights (.pt) file path (required)')
    parser.add_argument('-s', '--size', nargs='+', type=int, default=[640], help='Inference size [H,W] (default [640])')
    parser.add_argument('--p6', action='store_true', help='P6 model')
    parser.add_argument('--opset', type=int, default=17, help='ONNX opset version')
    parser.add_argument('--simplify', action='store_true', help='ONNX simplify model')
    parser.add_argument('--dynamic', action='store_true', help='Dynamic batch-size')
    parser.add_argument('--batch', type=int, default=1, help='Static batch-size')
    parser.add_argument('--split_outputs', action='store_true', help='Split outputs into boxes, scores, and classes')
    args = parser.parse_args()

    if not os.path.isfile(args.weights):
        raise SystemExit('Invalid weights file')
    if args.dynamic and args.batch > 1:
        raise SystemExit('Cannot set dynamic batch-size and static batch-size at same time')
    return args


if __name__ == '__main__':
    args = parse_args()
    main(args)
then running python3 export_yoloV5.py --weights best.pt --size 640 --simplify --dynamic --opset 17 --split_outputs the resulting .onnx file works for me in terms of getting the stream/video but it doesnt infere on the stream e.g no bounding boxes,confidance or classes.

**Comment by marcoslucianops** [comment]
@Sanelembuli98 it's not the same issue.

**Comment by Sanelembuli98** [comment]
@marcoslucianops its sounds like the same problem to me. What would be the solution to this one?

**Comment by DanielSACE** [comment]
> Added YOLO11

Hello, I used YOLO11,but  the .engine model still doesn't detect anything. Can you help me ansys it?

 WARNING: Deserialize engine failed because file path: /home/ssa/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine open error
0:00:02.621228672 118779 0xaaaaed9978f0 WARN                 nvinfer gstnvinfer.cpp:677:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1897> [UID = 1]: deserialize engine from file :/home/ssa/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine failed
0:00:02.824899968 118779 0xaaaaed9978f0 WARN                 nvinfer gstnvinfer.cpp:677:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:2002> [UID = 1]: deserialize backend context from engine from file :/home/ssa/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine failed, try rebuild
0:00:02.824951104 118779 0xaaaaed9978f0 INFO                 nvinfer gstnvinfer.cpp:680:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1923> [UID = 1]: Trying to create engine from model files
WARNING: [TRT]: onnx2trt_utils.cpp:375: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
WARNING: [TRT]: onnx2trt_utils.cpp:403: One or more weights outside the range of INT32 was clamped
WARNING: [TRT]: Tensor DataType is determined at build time for tensors not marked as input or output.

Building the TensorRT Engine

WARNING: [TRT]: Tactic Device request: 8400MB Available: 7352MB. Device memory is insufficient to use tactic.
WARNING: [TRT]: Skipping tactic 3 due to insufficient memory on requested size of 8400 detected for tactic 0x0000000000000004.
Try decreasing the workspace size with IBuilderConfig::setMemoryPoolLimit().
WARNING: [TRT]: Tactic Device request: 8400MB Available: 7352MB. Device memory is insufficient to use tactic.
WARNING: [TRT]: Skipping tactic 8 due to insufficient memory on requested size of 8400 detected for tactic 0x000000000000003c.
Try decreasing the workspace size with IBuilderConfig::setMemoryPoolLimit().
WARNING: [TRT]: Tactic Device request: 8400MB Available: 7352MB. Device memory is insufficient to use tactic.
WARNING: [TRT]: Skipping tactic 13 due to insufficient memory on requested size of 8400 detected for tactic 0x0000000000000074.
Try decreasing the workspace size with IBuilderConfig::setMemoryPoolLimit().
Building complete

0:04:38.748967936 118779 0xaaaaed9978f0 INFO                 nvinfer gstnvinfer.cpp:680:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1955> [UID = 1]: serialize cuda engine to file: /home/ssa/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine successfully
WARNING: [TRT]: The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
INFO: [Implicit Engine Info]: layers num: 2
0   INPUT  kFLOAT input           3x640x640       
1   OUTPUT kFLOAT output          8400x6          

0:04:38.988789056 118779 0xaaaaed9978f0 INFO                 nvinfer gstnvinfer_impl.cpp:328:notifyLoadModelStatus:<primary_gie> [UID 1]: Load new model:/home/ssa/DeepStream-Yolo-master/config_infer_primary_yolo11.txt sucessfully

Runtime commands:
	h: Print this help
	q: Quit

	p: Pause
	r: Resume

NOTE: To expand a source in the 2D tiled display and view object details, left-click on the source.
      To go back to the tiled display, right-click anywhere on the window.


**PERF:  FPS 0 (Avg)	
**PERF:  0.00 (0.00)	
** INFO: <bus_callback:239>: Pipeline ready

Opening in BLOCKING MODE 
NvMMLiteOpen : Block : BlockType = 279 
NVMEDIA: Reading vendor.tegra.display-size : status: 6 
NvMMLiteBlockCreate : Block : BlockType = 279 
** INFO: <bus_callback:225>: Pipeline running

**PERF:  40.92 (40.48)	
**PERF:  40.80 (40.84)	
**PERF:  41.16 (40.83)	
**PERF:  41.12 (40.97)	
**PERF:  40.22 (40.81)	
**PERF:  40.36 (40.74)	
**PERF:  41.36 (40.84)	
**PERF:  40.57 (40.81)	
**PERF:  41.11 (40.83)	
**PERF:  40.50 (40.81)	
**PERF:  40.38 (40.77)	
nvstreammux: Successfully handled EOS for source_id=0
** INFO: <bus_callback:262>: Received EOS. Exiting ...

Quitting
App run successful

**Comment by DanielSACE** [comment]
@marcoslucianops can you give me some advice?

**Comment by marcoslucianops** [comment]
https://github.com/marcoslucianops/DeepStream-Yolo/issues/631
